### ISSUE TITLE
chap04_Classification_gradientTape_sigmoid

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#36]
"""
gradientTape + sigmoid
"""
import tensorflow as tf # ver2.x
tf.executing_eagerly() # True

### 1. input/output 변수 정의 
inputs = tf.Variable([[1., 2.], [2, 3], [3, 1], [4, 3], [5, 3], [6, 2]])
-[1., 2.] 원소 하나만 실수로 바꿔주면 나머지도 실수로 바꿔진다.
outputs = tf.Variable([[0.], [0], [0], [1], [1], [1]]) 

### 2. model : model 클래스
class Model(tf.keras.Model) : # tf.karas.Model 이용해서 자식클래스(부모클래스) 정의
    def __init__(self) : # 생성자 
        super().__init__() # 부모생성자 호출
        self.W = tf.Variable(tf.random.normal([2,1])) # 기울기[입력수, 출력수]
        self.B = tf.Variable(tf.random.normal([1])) # 절편[출력수]
    def call(self, inputs) : # 메서드 재정의
        return tf.matmul(inputs,self.W) + self.B # 회귀방정식(예측치)

-요런식으로 5번에서 모델 만들 것이야.
-model = Model()
-model.call(inputs) : call은 생략가능
-model(inputs)

### 3. 손실 함수
def loss(model, inputs, outputs) :
    sigmoid = tf.sigmoid(model(inputs))
    return -tf.reduce_mean(outputs * tf.math.log(sigmoid) + (1-outputs) * tf.math.log(1-sigmoid)) # cross entropy

### 4. 미분계수(기울기) 계산  
def gradient(model, inputs, outputs) : 
    with tf.GradientTape() as tape:
        loss_value = loss(model, inputs, outputs) # 손실함수 호출  
        grad = tape.gradient(loss_value, [model.W, model.B])
        # 미분계수 -> 기울기와 절편 업데이트
    return grad # 업데이트 결과 반환

### 5. model 생성
model = Model() # 생성자

### 6. model 최적화 객체
opt = tf.keras.optimizers.Adam(learning_rate = 0.01)

- 학습 전 초기손실값과 w,b
print("초기 손실값 : {:.6f}".format(loss(model, inputs, outputs)))
print("w : {}, b : {}".format(model.W.numpy(), model.B.numpy()))
'''
초기 손실값 : 1.336757
w : [[-0.57089543]
 [ 1.6738539 ]], b : [-0.17710568]
'''

### 7. 반복학습
for step in range(300) :
    grad = gradient(model, inputs, outputs) # 기울기 계산
    - 기울기 -> 최적화 객체 반영
    opt.apply_gradients(zip(grad, [model.W, model.B]))
    
    if (step+1) % 20 ==0 :
        print("step = {}, loss = {:.6f}".format((step+1), loss(model, inputs, outputs)))

- model 최적화
print("최종 손실값 : {:.6f}".format(loss(model, inputs, outputs)))
print("w : {}, b : {}".format(model.W.numpy(), model.B.numpy()))
'''
최종 손실값 : 0.644834
w : [[0.03869871]
 [0.5339896 ]], b : [-0.74212307]
'''
'''
sgd -> adam
최종 손실값 : 0.555574
w : [[0.18642066]
 [0.18759048]], b : [-0.71428424]
'''

- model test
sigmoid = tf.sigmoid(model.call(inputs)) # 0~1 확률
pred = tf.cast(sigmoid > 0.5, tf.float32) # 관계식 -> 1 or 0

print(pred.numpy())
'''
tf.Tensor(
[[0.]
 [0.]
 [1.]
 [1.]
 [1.]
 [1.]], shape=(6, 1), dtype=float32)
'''
print(outputs.numpy())
'''
<tf.Variable 'Variable:0' shape=(6, 1) dtype=float32, numpy=
array([[0.],
       [0.],
       [0.],
       [1.],
       [1.],
       [1.]], dtype=float32)>
'''

- 불필요한 차원 축소할 때
y_true = tf.squeeze(outputs) # 2차원 -> 1차원
y_pred = tf.squeeze(pred)
print(y_true.numpy()) # [0. 0. 0. 1. 1. 1.]
print(y_pred.numpy()) # [0. 0. 1. 1. 1. 1.]
